runtime/src/main/java/org/corfudb/runtime/view/stream/StreamAddressSpace.java

```diff
@@ -129,81 +120,137 @@ public long getTail() {
             return;
         }
 
-        // Note: if a negative value is passed to this API the cardinality
-        // of the bitmap is returned, which would be incorrect as we would
-        // be removing all addresses upon an invalid trim mark.
-        long numAddressesToTrim = bitmap.rankLong(trimMark);
+        // At this point point the trim mark is positive and we need to compute
+        // the rank. The rank will indicate how many addresses need to be trimmed
+        long rank = bitmap.rankLong(newTrimMark);
 
-        if (numAddressesToTrim <= 0) {
+        if (rank <= 0) {
             return;
         }
 
-        List<Long> addressesToTrim = new ArrayList<>();
-        LongIterator it = bitmap.getLongIterator();
-        for (int i = 0; i < numAddressesToTrim; i++) {
-            addressesToTrim.add(it.next());
+        Roaring64Bitmap trimmedBitmap = new Roaring64Bitmap();
+        LongIterator iterator = this.bitmap.getReverseLongIterator();
+        while (iterator.hasNext()) {
+            long current = iterator.next();
+
+            if (current <= newTrimMark) {
+                log.info("trim: stream trim mark moved from {} to {}", this.trimMark, current);
+                this.trimMark = current;
+                break;
+            }
+
+            trimmedBitmap.addLong(current);
         }
 
-        log.trace("trim: Remove {} addresses for trim mark {}", addressesToTrim.size(), trimMark);
+        this.bitmap = trimmedBitmap;
+
+        if (log.isTraceEnabled()) {
+            log.trace("trim: trimMark={} streamTrimMark {} size {} addresses={}", newTrimMark,
+                    this.trimMark,
+                    this.bitmap.getLongCardinality(),
+                    this.bitmap.toArray());
+        }
+    }
 
-        // Remove and set trim mark
-        if (!addressesToTrim.isEmpty()) {
-            removeAddresses(addressesToTrim);
+    /**
+     * Add range helper method that validates the range before adding it
+     * to the bitmap, this is to protect against this bug: https://github.com/RoaringBitmap/RoaringBitmap/pull/445
+     * @param toAdd bitmap to add the range to
+     * @param start start address (inclusive)
+     * @param end end address (exclusive)
+     */
+    private void addRangeHelper(Roaring64Bitmap toAdd, long start, long end) {
+        if (end == 0 || end <= start) {
+            throw new IllegalArgumentException("Invalid range [" + start + "," + end + ")");
         }
+        toAdd.add(start, end);
     }
 
     /**
      * Get addresses in range (end, start], where start > end.
      *
      * @return Bitmap with addresses in this range.
      */
-    public StreamAddressSpace getAddressesInRange(StreamAddressRange range) {
-        Roaring64NavigableMap addressesInRange = new Roaring64NavigableMap();
+    public StreamAddressSpace getAddressesInRange2(StreamAddressRange range) {
+        Roaring64Bitmap addressesInRange = new Roaring64Bitmap();
```

Comment:
Remove this.

Reply:
getAddressesInRange2 can be removed, right?